frame/bags-list/src/list/mod.rs

```diff
@@ -384,6 +384,77 @@ impl<T: Config> List<T> {
 		})
 	}
 
+	/// Put `heavier_id` to the position directly in front of `lighter_id`. Both ids must be in
+	/// same bag  and the `weight_of` `lighter_id` must be less than that of `heavier_id`
+	pub(crate) fn put_in_front_of(
+		lighter_id: &T::AccountId,
+		heavier_id: &T::AccountId,
+		weight_of: Box<dyn Fn(&T::AccountId) -> VoteWeight>,
+	) -> Result<(), crate::pallet::Error<T>> {
+		use crate::pallet;
+		use frame_support::ensure;
+
+		let lighter_node = Node::<T>::get(&lighter_id).ok_or(pallet::Error::IdNotFound)?;
+		let mut heavier_node = Node::<T>::get(&heavier_id).ok_or(pallet::Error::IdNotFound)?;
+
+		ensure!(lighter_node.bag_upper == heavier_node.bag_upper, pallet::Error::NotInSameBag);
+		ensure!(
+			heavier_node.next.as_ref() != Some(lighter_id),
+			pallet::Error::AlreadyHigherPosition
+		);
+
+		// this is the most expensive check, so we do it last
+		ensure!(weight_of(&heavier_id) > weight_of(&lighter_id), pallet::Error::NotHeavier);
+
+		// we need to write the bag to storage if its head and/or tail is updated
+		if lighter_node.is_terminal() || heavier_node.is_terminal() {
+			let mut bag = Bag::<T>::get(lighter_node.bag_upper).ok_or_else(|| {
+				debug_assert!(false, "bag that should exist cannot be found");
+				crate::log!(warn, "bag that should exist cannot be found");
+				pallet::Error::BagNotFound
+			})?;
+			debug_assert!(bag.iter().count() > 1);
+
+			if bag.head.as_ref() == Some(lighter_id) {
+				debug_assert!(lighter_node.next.is_some(), "lighter node must have next if head");
+				bag.head = Some(heavier_id.clone());
+			}
+
+			// re-assign bag tail if lighter is the tail
+			if bag.tail.as_ref() == Some(heavier_id) {
+				debug_assert!(heavier_node.prev.is_some(), "heavier node must have prev if tail");
+				bag.tail = heavier_node.prev.clone();
+			}
+
+			// within this block we know the bag must have been updated, so we update storage
+			bag.put()
+		}
+
+		// cut heavier out of the list, updating its neighbors
+		heavier_node.excise();
+
+		// re-fetch `lighter_node` from storage since it may have been updated when heavier node was
+		// excised
+		let mut lighter_node = Node::<T>::get(&lighter_id).ok_or(pallet::Error::IdNotFound)?; // TODO this should never fail, but if does could be bad news
```

Comment:
This should never fail; should we make it an expect?

Reply:
yup